### PR TITLE
Support (configurable) exponential backoff

### PR DESCRIPTION
**Pull request by sperand-io**
This PR does a few thing:

- allows library user to explicitly override default behaviors for timeouts, retry behavior, and queue size
- makes the default retry delay logic a bit smarter, factoring in jitter if passed in as an option and respecting configurable factor (using logic from [segmentio/backo](https://github.com/segmentio/backo))
- respects new `maxItems` option, removing still-queued tasks oldest first
- respects new `maxAttempts` option, discarding after exceeded

All defaults are maintained; this is a nonbreaking change.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -29,17 +29,26 @@ function bind(func, obj) {
...
   this.fn = fn;
+  this.maxItems = opts.maxItems || Infinity;
+  this.maxAttempts = opts.maxAttempts || Infinity;
 
   this.timeouts = {
-    ACK_TIMER: 1000,
-    RECLAIM_TIMER: 3000,
-    RECLAIM_TIMEOUT: 10000,
-    RECLAIM_WAIT: 500,
-    MAX_QUEUE_DELAY: 30000
```
Comment (f2prateek) [inline lib/index.js]: Removing this is a breaking change I think. Should be ok to do so if we don't use it in a.js-integration-segmentio.

File: lib/index.js [hunk 2]
```diff
@@ -29,17 +29,26 @@ function bind(func, obj) {
  * @param {String} name The name of the queue. Will be used to find abandoned queues and retry their items
  * @param {processFunc} fn The function to call in order to process an item added to the queue
  */
-function Queue(name, fn) {
```
Comment (f2prateek) [inline lib/index.js]: Removing an arg here might also be a breaking change?

File: test/store.test.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ describe('Store', function() {
 
   beforeEach(function() {
     engine.clear();
-    store = new Store('name', 'id', keys, engine);
```
Comment (f2prateek) [inline test/store.test.js]: Why was this removed?

File: lib/index.js (hunk 2 above)
Comment (sperand-io) [inline lib/index.js]: we added an optional one, didnt remove one.

File: lib/index.js (hunk 1 above)
Comment (sperand-io) [inline lib/index.js]: ah yea— it's been moved to `MAX_RETRY_DELAY` in the `backoff` dictionary (with default value preserved). you're right though: if anyone was accessing it directly (wasn't configurable via argument before) then that would break.

I'm inclined to move forward anyway. To your point, it's not something we access directly in ajs

File: test/store.test.js (hunk 3 above)
Comment (sperand-io) [inline test/store.test.js]: realized it was just redundant :)

File: README.md [hunk 4]
```diff
@@ -30,6 +32,34 @@ queue.on('processed', function(err, res, item) {
 queue.start();
 ```
 
+### Options
+
+The queue can be initialized with the following options (*defaults shown*):
+
+```js
+var options = {
+  ackTimer: 1000,        // ack interval in ms
```
Comment (tsholmes) [inline README.md]: The ack/reclaim settings are pretty advanced settings that are nearly impossible to get right. I'm honestly not sure why I made them configurable in the first place, but we probably shouldn't be suggesting people can modify them.

Comment (sperand-io) [inline README.md]: cool, i'll hide those and just expose the backoff stuff
